### PR TITLE
Support fnmatch pattern in SFTP Hook and Sensor

### PR DESCRIPTION
**Pull request by saveriogzz**
Effort to give users the possibility to just operates on specific files using the SFTP provider. Please refer to the old PR below for details.

related
- issue: #15332
- PR by @blcksrx : #15409
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/sftp/hooks/sftp.py [hunk 1]
```diff
@@ -212,7 +218,9 @@ def delete_directory(self, path: str) -> None:
         conn = self.get_conn()
         conn.rmdir(path)
 
-    def retrieve_file(self, remote_full_path: str, local_full_path: str) -> None:
+    def retrieve_file(
+        self, remote_full_path: str, local_full_path: str, fnmatch_pattern: Optional[str] = None
+    ) -> None:
```
Comment (saveriogzz) [inline airflow/providers/sftp/hooks/sftp.py]: @uranusjr, in #15409, already mentioned that this implementation could cause headaches to users inasmuch "it subtly changes semantics depending on the arguments". We can start from here and think about a better solution.

**Comment by uranusjr** [comment]
I think my https://github.com/apache/airflow/pull/15409#discussion_r616309811 still applies. `fnmatch` is indeed a better solution, but we need to design the functions better. Mayne leave `retrieve_file` along and implement a new `retrieve_files` instead?

File: airflow/providers/sftp/hooks/sftp.py [hunk 2]
```diff
@@ -250,14 +265,22 @@ def delete_file(self, path: str) -> None:
...
+        :param fnmatch_pattern: optional pattern to filter the remote_full_path files
+        :type: fnmatch_pattern: Optional[str]
         """
         conn = self.get_conn()
+        if fnmatch_pattern:
+            for file in conn.listdir(path):
+                if fnmatch(file, fnmatch_pattern):
+                    path = file
+                    break
+
```
Comment (saveriogzz) [inline airflow/providers/sftp/hooks/sftp.py]: @uranusjr , how about this method called by the [SFTPSensor](https://github.com/apache/airflow/blob/main/airflow/providers/sftp/sensors/sftp.py#L49)?
Should we create a different method to be able to catch all the files matching the pattern or we just wait for the the first of the list as it is implemented now?

File: airflow/providers/sftp/sensors/sftp.py [hunk 3]
```diff
@@ -57,13 +57,23 @@ def __init__(
...
+                return False
+        else:
+            self.log.info('Poking for files matching pattern %s into %s', self.fnmatch_pattern, self.path)
+            try:
+                mod_time = self.hook.get_mod_time_pattern(self.path, self.fnmatch_pattern)
+                self.log.info('Found Files last modified %s', str(mod_time))
+            except OSError as e:
+                if e.errno != SFTP_NO_SUCH_FILE:
+                    raise e
+                return False
```
Comment (saveriogzz) [inline airflow/providers/sftp/sensors/sftp.py]: here we have two options:
- not having specified `fnmatch_pattern`
- having specified it

In the first case, the parameter `path` is as usual the full path to the file.
In the second case, the `path` will be the path to the directory whilst `fnmatch_pattern` will be useful to catch all the files we are interested in.
In this second case, the new method [get_mod_time_pattern](https://github.com/saveriogzz/airflow/blob/AIRFLOW-15332/airflow/providers/sftp/hooks/sftp.py#L311), which returns a dictionaries with full paths to files as keys and their modification time as values.

File: airflow/providers/sftp/hooks/sftp.py [hunk 4]
```diff
@@ -265,25 +284,51 @@ def delete_file(self, path: str) -> None:
...
-        Returns modification time.
+        Removes files matching the pattern on the FTP Server
+
+        :param dir_path: full path to the remote directory
+        :type dir_path: str
+        :param fnmatch_pattern: pattern to filter the dir_path files
+        :type fnmatch_pattern: str
+        """
+        conn = self.get_conn()
+        for file in conn.listdir(dir_path):
```
Comment (saveriogzz) [inline airflow/providers/sftp/hooks/sftp.py:297]: this is wrong because it should raise errors in case no matches are found, at the moment it's not

**Comment by uranusjr** [comment]
It seems like your rebase went wrong.

**Review by saveriogzz — COMMENTED** [review]
> It seems like your rebase went wrong.

very wrong.... Sorry for this!
Should I close this PR and create a cleaner one?

**Comment by saveriogzz** [comment]
I'm going ahead closing this because I realized that the contribution is far from being ready to merge.

**Pull request closed without merging**